lib: add chainErrors API to avoid error swallowing

aduh95 opened the pull request:
I think the current behaviour is wrong:

- it waits for the file to be closed, but there's no need for this, the error could be passed directly to the user.
- It swallows the error on closing. Failing to close a file is fairly rare in practice, but I think it would make sense to use the default behaviour added in https://github.com/nodejs/node/pull/37174 (rethrow the error if one occurs).

~Conservatively marking this as semver-major as it might raise an error where it didn't before.~
EDIT: I've repurposed this PR to aggregate errors instead.

/cc @nodejs/fs 

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/master/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

RaisinTen reviewed (approved):
The current behaviour is indeed incorrect.
This was introduced in this commit: https://github.com/nodejs/node/commit/ef55324f1a572d410e8022b5152ebe4778e387e0#
Before the commit, it worked just like your patch:
https://github.com/nodejs/node/blob/662fdfed378db3fde5de4fde09e8a8bfbebcac4d/src/node.js#L621-L625

ronag reviewed (dismissed):
I think this is better?

```js
      if (isUserFd) {
        callback(writeErr);
      } else {
        fs.close(fd, function close(er) {
          callback(err || writeErr);
        });
      }
```

aduh95 commented:
> I think this is better?
> 
> ```js
>       if (isUserFd) {
>         callback(writeErr);
>       } else {
>         fs.close(fd, function close(er) {
>           callback(err || writeErr);
>         });
>       }
> ```

This would swallow the `writeErr`, which is worse than the current behavior IMHO.

ronag commented:
> This would swallow the writeErr, which is worse than the current behavior IMHO.

Swallow how? We always propagate the most inner failure.

aduh95 commented:
> Swallow how? Only if `fs.close` fails (which is very unlikely). We always propagate the most inner failure.

You're right, I think I misread your suggestion at first, sorry.

Why do you think it is better? Is it because it's more consistent with other fs functions?
I'd say it's not useful to wait for the file to close to give the error back to the user, as they would more likely care for a write error than a close error (which is probably why the current way has been implemented in the first place).

ronag commented:
> Why do you think it is better? Is it because it's more consistent with other fs functions?
I'd say it's not useful to wait for the file to close to give the error back to the user, as they would more likely care for a write error than a close error (which is probably why the current way has been implemented in the first place).

Consistency. We always propagate innermost error.

aduh95 commented:
Hum it seems we are consistently ignoring error on close, and prefer report the other error:

https://github.com/nodejs/node/blob/ffa874fd4c6a0c265c5b1e96ead97e4e233d6d67/lib/fs.js#L1286-L1292
https://github.com/nodejs/node/blob/ffa874fd4c6a0c265c5b1e96ead97e4e233d6d67/lib/fs.js#L822-L824

I think that makes sense as the close error is less interesting when you are performing a write operation.

BridgeAR commented:
@ronag 

> Consistency. We always propagate innermost error.

This makes debugging difficult. Instead, we should always propagate the most outer error. The inner error is likely an error that is related to the former one without containing the most important information.

We could combine the errors by attaching inner errors to the outer one.

ronag commented:
> This makes debugging difficult. Instead, we should always propagate the most outer error. The inner error is likely an error that is related to the former one without containing the most important information.

This is not how we do it in a lot of other places. Whether we propagate the inner or outer error is a bigger discussion I think we should have (@mcollina, @lpinca, @jasnell, @addaleax) maybe in a separate issue?

My current understanding is that we always propagate the inner error. Regardless whether that is the best thing or not, I think it's important we are consistent, i.e. propagate the inner error for now until we have a different consensus.

aduh95 commented:
Chaining errors would work great as long as we are able to generate comprehensible stack traces. Something like a `causedBy` property on the error object, maybe?

mcollina reviewed (changes requested):
This needs a test.

aduh95 commented:
> This needs a test.

@mcollina Not sure how to test this as it only occurs if both write and close operations fail. Can you guide me on how to make `fs.close` reliably fail?

mcollina commented:
Maybe @jasnell have some ideas?

ronag commented:
Monkey patch fs?

aduh95 commented:
I've taken the suggestion from BridgeAR to chain errors instead of trying to guess which one should be reported to the user. I've also added tests, PTAL.

ronag commented on lib/internal/errors.js:
```diff
@@ -136,6 +139,14 @@ const maybeOverridePrepareStackTrace = (globalThis, error, trace) => {
+    return error;
+  }
+  return originalError;
+};
```
Should this be: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AggregateError

ronag commented on lib/_http_client.js:
```diff
@@ -789,14 +789,14 @@ function onSocketNT(req, socket, err) {
-        finished(socket.destroy(err || req[kError]), (er) => {
-          _destroy(req, er || err);
+        finished(socket.destroy(chainErrors(req[kError], err)), (er) => {
+          _destroy(req, chainErrors(err, er));
```
Please don't change stream related stuff before changing the core streams.

aduh95 replied:
Not sure what you mean by `changing the core streams`, so I've removed the changes from the PR.

ronag commented on lib/_http_client.js:
```diff
@@ -789,14 +789,14 @@ function onSocketNT(req, socket, err) {
     }
 
-    _destroy(req, err || req[kError]);
+    _destroy(req, chainErrors(req[kError], err));
```
Please don't change stream related stuff before changing the core streams.

ronag commented on lib/internal/fs/streams.js:
```diff
@@ -276,7 +279,7 @@ ReadStream.prototype._destroy = function(err, cb) {
   // any pending IO (kIsPerformingIO) to complete (kIoDone).
   if (this[kIsPerformingIO]) {
-    this.once(kIoDone, (er) => close(this, err || er, cb));
+    this.once(kIoDone, (er) => close(this, chainErrors(er, err), cb));
```
Please don't change stream related stuff before changing the core streams.

ronag commented on lib/internal/fs/streams.js:
```diff
@@ -437,7 +440,7 @@ WriteStream.prototype._destroy = function(err, cb) {
   // any pending IO (kIsPerformingIO) to complete (kIoDone).
   if (this[kIsPerformingIO]) {
-    this.once(kIoDone, (er) => close(this, err || er, cb));
+    this.once(kIoDone, (er) => close(this, chainErrors(er, err), cb));
```
Please don't change stream related stuff before changing the core streams.

mcollina reviewed (changes requested):
As @ronag said in a commit, this should introduce https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AggregateError if it's not available.

aduh95 commented:
I've made changes to use `AggregateError`. The stack trace is a bit less helpful, but I think this is a V8 issue (it's a bit weird they have a better support for the unofficial `cause` property than the standard `AggregateError`, surely it'll be fixed in a future V8 update).

ronag reviewed (commented):
RSLGTM

mcollina commented on lib/internal/streams/destroy.js:
```diff
@@ -56,7 +59,7 @@ function destroy(err, cb) {
   if (!s.constructed) {
     this.once(kDestroy, function(er) {
-      _destroy(this, err || er, cb);
+      _destroy(this, chainErrors(er, err), cb);
```
please use AggregateError directly.

aduh95 replied:
I disagree, I think the current implementation is more user friendly: if there's only one error, we should return it to the user without wrapping it into an `AggregateError`.

mcollina replied:
I mean to not use chainErrors.

aduh95 replied:
Do you mean you prefer to inline it? I'd rather not, but if you think it's the way to go, let's do that.
```suggestion
      _destroy(this, er && err ? new AggregateError([er, err]) : (err || er), cb);
```

aduh95 replied:
Also, do you mean for this specific change or all the changes in this PR?

legendecas commented:
> Chaining errors would work great as long as we are able to generate comprehensible stack traces. Something like a `causedBy` property on the error object, maybe?

There is an active TC39 proposal seeking for adding error chaining support to ECMAScript: https://github.com/tc39/proposal-error-cause (currently stage 2).

While `AggregateError` somehow does "aggregate"s the errors and not losing tracking of them or swallow them, the `AggregateError` doesn't seem to be saying that these errors were correlated, i.e. they can be totally no relation to each other, like in the case of `Promise.any`.

ronag commented:
> There is an active TC39 proposal seeking for adding error chaining support to ECMAScript: https://github.com/tc39/proposal-error-cause (currently stage 2).

Nice! I'm personally fine with implementing this before it's part of spec.

mcollina commented:
I'm -1 in adopting a Stage 2 proposal in anything but code clearly marked as experimental.

ronag commented:
I'm -0 on adopting `AggregateError` for this only to later again change it to an `Error` with `cause`. Can we wait a little until the proposal is further along? @mcollina Would you be fine with stage 3 or do you want it fully part of spec?

mcollina commented:
We shipped several stage 3 proposals in the past.

aduh95 commented:
@mcollina are you fine with keeping `chainErrors` function as is (using `AggregateError` for now) so it’s easy to switch to the proposal-error-cause once it reaches stage 3?

aduh95 commented:
After giving it some more thoughts, I don't think it's a good use case for using `cause` here: those errors are more often than not not related (`fs.close` don't fail because the write operation failed, although both errors could have a common cause).

I've changed the function name to be clear that we are not chaining the errors and added some more tests, PTAL.

RaisinTen commented on test/message/error_chainErrors.js:
```diff
@@ -0,0 +1,10 @@
+'use strict';
+
+require('../common');
+const { chainErrors } = require('internal/errors');
```
```suggestion
const { aggregateTwoErrors } = require('internal/errors');
```
Should this file be named `test/message/error_aggregateTwoErrors.js` instead?

RaisinTen commented on test/message/error_chainErrors.js:
```diff
@@ -0,0 +1,10 @@
+const originalError = new Error('original');
+const err = new Error('second error');
+
+throw chainErrors(err, originalError);
```
```suggestion
throw aggregateTwoErrors(err, originalError);
```

RaisinTen commented on test/message/error_chainErrors.out:
```diff
@@ -0,0 +1,11 @@
+*/error_chainErrors.js:*
+throw chainErrors(err, originalError);
+^
+AggregateError: original
+    at Object.<anonymous> (*test*message*error_chainErrors.js:*:*)
```
```suggestion
*/error_aggregateTwoErrors.js:*
throw aggregateTwoErrors(err, originalError);
^
AggregateError: original
    at Object.<anonymous> (*test*message*error_aggregateTwoErrors.js:*:*)
```
New filename: `test/message/error_aggregateTwoErrors.out`

RaisinTen commented on test/parallel/test-error-chainErrors.js:
```diff
@@ -0,0 +1,46 @@
+// Flags: --expose-internals
```
Filename: `test/parallel/test-error-aggregateTwoErrors.js`

mcollina reviewed (approved):
lgtm

ronag commented:
This is very semver-major IMO.

ronag reviewed (changes requested):
Putting as request changes until semver has been discussed.

aduh95 commented:
> This is very semver-major IMO. Code that checks for e.g. error codes will stop working.

Should I try to copy `code` and `name` property of the outer error on the `AggregateError` to mimic the current behavior?

ronag commented:
> Should I try to copy code and name property of the outer error on the AggregateError to mimic the current behavior?

Feels a bit hacky and I'm not sure that's sufficient. Sometimes user look at message.

aduh95 commented:
> Sometimes users look at message.

FWIW the message is already copied from the most outer error in the current implementation.

ronag commented:
> FWIW the message is already copied from the most outer error in the current implementation.

That also feels a little hacky :)

aduh95 commented:
I've added documentation for the impacted `fs` methods, PTAL.

aduh95 commented:
Ping @ronag

ronag commented:
I don't have time at the moment to put much thought into this. I'd just like a confirmation that those that have approved the changes (@jasnell, @mcollina, @addaleax) have considered that existing (and future) user code might break or get confused by getting different errors types in an unpredictable manner. Before there was at least some predictability in the sense that the inner error always propagates. This has some significant breaking impact potential. Especially code that check for error codes.

If this has been taken into consideration then I will defer to the other reviewers and won't have any further objections.

ronag commented:
@mcollina Would we apply the same logic to streams?

legendecas commented:
> @mcollina After giving it some more thoughts, I don't think it's a good use case for using cause here: those errors are more often than not related (fs.close don't fail because the write operation failed, although both errors could have a common cause).

I'm confusing in this case, if the `fs.close` is a logically subsequent operation of the failure of `fs.write`, why are those errors not related? The `cause` doesn't mean that those errors have to be a wrapper around the another, but they are triggered in a logical chain.

aduh95 commented:
> if the `fs.close` is a logically subsequent operation of the failure of `fs.write`

That's a big if, why do you think it would be the case? In the most case, `fs.write` fails without causing a `fs.close` error, which leads me too think there is no logical implication between the two errors.

mcollina reviewed (changes requested):
After thinking about this for a while, I'm starting to think we should handle the `.code` property somehow. This is going to break almost all error handling cases that checked `.code`, making it significantly more complex and error prone.

mcollina reviewed (approved):
lgtm

aduh95 commented:
Landed in 104dac79ccd0
